File: leetcode/algorithms/string/remove-all-adjacent-duplicates-in-string.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Remove all Adjacent Duplicates in String"
*/

class Solution {
    private String removeAdjacentDuplicates(String S){
        StringBuilder result = new StringBuilder("");
        for(int i=0;i<S.length();){
            int j=i+1;
            if(j<S.length() && S.charAt(i)==S.charAt(j)){
                i+=2;
            } else {
                result.append(Character.toString(S.charAt(i)));
                ++i;
            }
        }
        return result.toString();
    }
    public String removeDuplicates(String S) {
        String lastString = S;
        while(true){
            String nextString = removeAdjacentDuplicates(lastString);
            if(nextString.equals(lastString)){
                break;
            }
            lastString=nextString;
        }
        return lastString;
        
    }
}

File: leetcode/algorithms/stack/remove-all-adjacent-duplicates-in-string.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Remove All Adjacent Duplicates in String"
*/

/**
 * @param {string} S
 * @return {string}
 */

function Stack(){
    this.stackArr = [];
}

Stack.prototype.push = function(val){
    this.stackArr.push(val);
}

Stack.prototype.pop = function(){
    if(this.stackArr.length===0){
        return null;
    }
    let value = this.stackArr.splice(this.stackArr.length-1,1);
    return value;
}

Stack.prototype.isEmpty = function(){
    return this.stackArr.length===0;
}

Stack.prototype.peek = function(){
    return this.stackArr.length===0 ? null : this.stackArr[this.stackArr.length-1];
}

var removeDuplicates = function(S) {
    
    const dupStack = new Stack();
    
    for(let i=0;i<S.length;++i){
        if(dupStack.isEmpty() || dupStack.peek()!=S[i]){
            dupStack.push(S[i]);
        } else {
            dupStack.pop();
        }
    }
    
    return dupStack.stackArr.join("");
    
};